modules/afamqp/afamqp.c

```diff
@@ -654,6 +684,31 @@ afamqp_worker_insert(LogThreadedDestDriver *s, LogMessage *msg)
   return LTR_SUCCESS;
 }
 
+static void
+_handle_heartbeat(void *cookie)
+{
+  AMQPDestDriver *self = (AMQPDestDriver *) cookie;
+
+  amqp_frame_t frame;
+  struct timeval tv = {0, 0};
+  gint status;
+  while (AMQP_STATUS_OK == (status = amqp_simple_wait_frame_noblock(self->conn, &frame, &tv)));
+  if (AMQP_STATUS_TIMEOUT != status)
```

Comment:
@gaborznagy
AMQP_STATUS_TIMEOUT is the right choice here. We want to repeatedly call the nonblocking read until either frames are available (AMQP_STATUS_OK), or until timeout of the nonblocking read (AMQP_STATUS_TIMEOUT). We need to signal error in case there was a frame error, hence the not-equal AMQP_STATUS_TIMEOUT check. Specially, we will signal error when the status is AMQP_STATUS_HEARTBEAT_TIMEOUT, and in other error conditions too.

Reply:
Reading back my comment, I did not write what I intended, obviously simply switching the 2 return values wouldn't work, sorry for that.
I see that we don't want to handle AMQP_STATUS_TIMEOUT as an error.
Most of the time this will be the return value in case of idle connections.

What I wanted to point out is that AMQP_STATUS_HEARTBEAT_TIMEOUT shouldn't be an unexpected error either. It's not the same as for example AMQP_STATUS_SOCKET_ERROR.
so not just the read/wait timeout is not an error, but neither heartbeat-timeout, nor the case of receiving a frame.

Reply:
In short: we should handle:
- read/wait timeout 
- heartbeat timeout
- frame received ("ok" status)
- other errors
differently.
Currently, only non-read/wait timeouts handled, everything else interpreted as an unexpected error.
Even `AMQP_STATUS_OK` would be handled as an error if a previous while loop wouldn't prevent reaching this statement.

Reply:
@gaborznagy
I think the heartbeat is used to detect broken connections. That's why it seemed reasonable to do a disconnect/reconnect.
Certainly, if there is a better way to handle this kind of error than disconnect/reconnect, let me know and I will update the patch according to it.

Reply:
I accept your reason.

Reply:
Thanks for explaining. :)